crabs/detection_tracking/datamodule.py

```diff
@@ -0,0 +1,241 @@
+    train_ids : list
+        List of image IDs for the training set.
+    test_ids : list
+        List of image IDs for the validation (or test) set.
+    coco_data : dict
+        Dictionary containing COCO dataset information.
+    """
+
+    def __init__(self, main_dir, annotation, batch_size, seed_n=42):
+        super().__init__()
+        self.main_dir = main_dir
+        self.annotation = annotation
+        self.batch_size = batch_size
+        self.seed_n = seed_n
+
+    def prepare_data(self):
+        pass
+
+    def setup(self, stage=None):
+        """
+        Sets up the data loader for the specified stage ('fit' or 'test').
+
+        Returns
+        ----------
+        None
+        """
+        with open(self.annotation) as json_file:
+            self.coco_data = json.load(json_file)
+            exclude_video_file_list = [
+                "09.08.2023-03-Left",
+                "10.08.2023-01-Left",
+                "10.08.2023-01-Right",
+            ]
+            all_ids = []
+
+            for image_info in self.coco_data["images"]:
+                image_id = image_info["id"]
+                image_file = image_info["file_name"]
+                video_file = image_file.split("_")[1]
+
+                if video_file not in exclude_video_file_list:
+                    all_ids.append(image_id)
+
+            if stage == "fit" or stage is None:
+                train_ids, _ = train_test_split(
+                    all_ids,
+                    train_size=0.8,
+                    shuffle=True,
+                    random_state=self.seed_n,
+                )
+                self.train_ids = train_ids
+            if stage == "test" or stage is None:
+                _, test_ids = train_test_split(
+                    all_ids,
+                    test_size=0.2,
+                    shuffle=True,
+                    random_state=self.seed_n,
+                )
+                self.test_ids = test_ids
+
+    def train_dataloader(self):
+        """
+        Returns the data loader for the validation (or test) set.
+
+        Returns
+        ----------
+        None
+        """
+        file_paths = [
+            next(
+                (
+                    info
+                    for info in self.coco_data["images"]
+                    if info["id"] == image_id
+                ),
+                None,
+            )["file_name"]
+            for image_id in self.train_ids
+        ]
+        train_dataset = myFasterRCNNDataset(
+            self.main_dir,
+            file_paths,
+            self.annotation,
+            transforms=get_train_transform(),
+        )
+        return DataLoader(
+            train_dataset,
+            batch_size=self.batch_size,
+            shuffle=True,
+            num_workers=4,
+            collate_fn=collate_fn,
+            persistent_workers=True,
+        )
+
+    def val_dataloader(self):
+        """
+        Returns the data loader for the evaluation set.
```

Comment:
During the training we want to shuffle the data, so that for every epoch the order of the data will be different so the network not going to cheat by any means having the data in order. Usually we don't shuffle during test so the results are consistent when we compared against different runs.
The same with persistent_workers, we need it to be true when dealing with big data. During test as we are dealing with smaller batch number, we don't really need it.